File: android/src/main/java/au/com/crazybean/mobilex/kurir/modules/contacts/ContactsViewModel.kt
package au.com.crazybean.mobilex.kurir.modules.contacts

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import au.com.crazybean.foundation.mvvm.ViewModel
import au.com.crazybean.mobilex.foundation.userdata.UserData
import au.com.crazybean.mobilex.kurir.data.kEmail
import au.com.crazybean.mobilex.kurir.data.model.User
import au.com.crazybean.mobilex.kurir.repository.contacts.ContactsRepository
import au.com.crazybean.mobilex.kurir.repository.users.UsersRepository

class ContactsViewModel(private val userData: UserData?,
                        private val usersRepository: UsersRepository?,
                        private val contactsRepository: ContactsRepository?) : ViewModel() {
    val contacts: LiveData<List<User>?>
        get() = MutableLiveData<List<User>?>().also { liveData ->
            val email = userData?.getString(kEmail)?: ""
            contactsRepository?.getContacts(email) { emails ->
                emails?.takeIf { it.isNotEmpty() }?.let { filters ->
                    usersRepository?.getUsers(filters) {
                        liveData.value = it
                    }
                }?: liveData.postValue(null)
            }
        }
}

File: ios/Kurir/modules/contacts/ContactsViewModel.swift
//
//  ContactsViewModel.swift
//  Kurir
//
//  Created by Loren on 19/9/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Crazybean
import Mobilex

class ContactsViewModel: ViewModel {
    private let userData: UserData?
    private let usersRepository: UsersRepository?
    private let contactsRepository: ContactsRepository?
    
    init(userData: UserData?, usersRepository: UsersRepository?, contactsRepository: ContactsRepository?) {
        self.userData = userData
        self.usersRepository = usersRepository
        self.contactsRepository = contactsRepository
    }
    
    var contacts: LiveData<[User]?> {
        get {
            return LiveData<[User]?>() { [weak self] liveData in
                if let email = self?.userData?.getString(forKey: KeysKt.kEmail, defaultValue: "") {
                    self?.contactsRepository?.getContacts(email: email, completion: { [weak self] emails in
                        if let emails = emails {
                            self?.usersRepository?.getUsers(emails: emails, completion: { users in
                                liveData.value = users
                            })
                        } else {
                            liveData.value = nil
                        }
                    })
                } else {
                    liveData.value = nil
                }
            }
        }
    }
}
